addons/promote/common/src/main/java/org/commonjava/indy/promote/validate/PromotionValidator.java

```diff
@@ -102,41 +118,85 @@ public ValidationRequest validate( PromoteRequest request, ValidationResult resu
+                                String ruleName =
+                                        new File( ruleRef ).getName(); // flatten in case some path fragment leaks in...
+
+                                ValidationRuleMapping rule = validationsManager.getRuleMappingNamed( ruleName );
+                                if ( rule != null )
                                 {
-                                    logger.debug( "{} succeeded", rule.getName() );
+                                    try
+                                    {
+                                        logger.debug( "Running promotion validation rule: {}", rule.getName() );
+                                        String error = rule.getRule().validate( req );
+                                        if ( StringUtils.isNotEmpty( error ) )
+                                        {
+                                            logger.debug( "{} failed", rule.getName() );
+                                            result.addValidatorError( rule.getName(), error );
+                                        }
+                                        else
+                                        {
+                                            logger.debug( "{} succeeded", rule.getName() );
+                                        }
+                                    }
+                                    catch ( Exception e )
+                                    {
+                                        if ( e instanceof PromotionValidationException )
+                                        {
+                                            exceptionHolder.set( (PromotionValidationException) e );
+                                        }
+
+                                        exceptionHolder.set( new PromotionValidationException(
+                                                "Failed to run validation rule: {} for request: {}. Reason: {}", e,
+                                                rule.getName(), request, e ) );
+                                    }
                                 }
                             }
-                            catch ( Exception e )
+                            finally
                             {
-                                if ( e instanceof PromotionValidationException )
-                                {
-                                    throw (PromotionValidationException) e;
-                                }
-
-                                throw new PromotionValidationException(
-                                        "Failed to run validation rule: {} for request: {}. Reason: {}", e,
-                                        rule.getName(), request, e );
+                                latch.countDown();
                             }
+                        } );
+                    }
+                    int timeout = 0;
+                    final int LATCH_INTERVAL = 1;
```

Comment:
This needs to be higher or we're going to burn a lot of CPU. Probably not a huge deal, but setting it to 100ms won't hurt anything.

Actually, if we have other code handling the promotion lock timeout in this way, that would explain the jump in CPU usage I noticed in production...

Reply:
Yep, sorry I missed that